### PR TITLE
Fix batch_norm abort on inputs with fewer than 2 dims

### LINKED ISSUE #186212: C++ API: at::batch_norm crashes with an internal assertion on invalid inputs
### 🐛 Describe the bug

Calling the C++ API `at::batch_norm` with invalid inputs causes the process to abort with an internal assertion instead of reporting a user-facing error.

### Reproducer

```cpp
#include <torch/torch.h>

int main() {
  torch::set_num_threads(1);

 torch::Tensor input =
      torch::randn({1}, torch::TensorOptions().dtype(torch::kFloat32));

  c10::optional<torch::Tensor> weight = c10::nullopt;
  c10::optional<torch::Tensor> bias = c10::nullopt;
  c10::optional<torch::Tensor> running_mean = c10::nullopt;
  c10::optional<torch::Tensor> running_var = c10::nullopt;

  at::batch_norm(
      input,
      weight,
      bias,
      running_mean,
      running_var,
      true,
      0.5,
      1.0 / 3.0,
      true);

  return 0;
}
```

### Observed behavior

The process aborts with:

```text
/root/pytorch/c10/util/SmallVector.h:333:
reference c10::SmallVectorTemplateCommon<long>::operator[](size_type) [T = long]:
Assertion `idx < size()' failed.

Aborted (core dumped)
```

Backtrace:

```text
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140734560691712) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140734560691712) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140734560691712, signo=signo@entry=6) at ./nptl/pthread_kill.c:89
#3  0x00007fff58c4d476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007fff58c337f3 in __GI_abort () at ./stdlib/abort.c:79
#5  0x00007fff58c3371b in __assert_fail_base (fmt=0x7fff58de8130 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=0x7fffb1515c80 <str> "idx < size()", file=0x7fffb15124da "/root/pytorch/c10/util/SmallVector.h", 
    line=333, function=<optimized out>) at ./assert/assert.c:94
#6  0x00007fff58c44e96 in __GI___assert_fail (assertion=0x7fffb1515c80 <str> "idx < size()", 
    file=0x7fffb15124da "/root/pytorch/c10/util/SmallVector.h", line=333, 
    function=0x7fffb1515ca0 <__PRETTY_FUNCTION__._ZN3c1025SmallVectorTemplateCommonIlvEixEm> "reference c10::SmallVectorTemplateCommon<long>::operator[](size_type) [T = long]") at ./assert/assert.c:103
#7  0x00007fff77c84530 in c10::SmallVectorTemplateCommon<long, void>::operator[] (this=0x7fff4f771220, idx=0)
    at /root/pytorch/c10/util/SmallVector.h:333
#8  0x00007fff7a99be5c in at::native::batch_norm_cpu (self=..., weight_opt=..., bias_opt=..., running_mean_opt=..., 
    running_var_opt=..., train=true, momentum=0.5, eps=0.33333333333333331)
    at /root/pytorch/aten/src/ATen/native/Normalization.cpp:879
#9  0x00007fff824cf59f in at::(anonymous namespace)::(anonymous namespace)::wrapper_CPU__native_batch_norm (input=..., 
    weight=..., bias=..., running_mean=..., running_var=..., training=true, momentum=0.5, eps=0.33333333333333331)
    at aten/src/ATen/RegisterCPU_1.cpp:3621
```

### Expected behavior

PyTorch should reject the invalid input and raise a user-facing exception instead of terminating the process with an internal assertion.

### Versions

PyTorch version: 2.12.0a0+git0d62256
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: linux (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.11.11 (main, Dec 11 2024, 16:28:39) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-110-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  32
On-line CPU(s) list:                     0-31
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Silver 4216 CPU @ 2.10GHz
CPU family:                              6
Model:                                   85
Thread(s) per core:                      1
Core(s) per socket:                      32
Socket(s):                               1
Stepping:                                7
BogoMIPS:                                4190.15
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat vnmi umip pku ospke avx512_vnni md_clear arch_capabilities
Virtualization:                          VT-x
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               1 MiB (32 instances)
L1i cache:                               1 MiB (32 instances)
L2 cache:                                128 MiB (32 instances)
L3 cache:                                16 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-31
Vulnerability Gather data sampling:      Unknown: Dependent on hypervisor status
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Mitigation; TSX disabled
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] intel-cmplr-lib-ur==2026.0.0
[pip3] intel-openmp==2026.0.0
[pip3] mkl-include==2026.0.0
[pip3] mkl-static==2026.0.0
[pip3] numpy==1.26.4
[pip3] onemkl-license==2026.0.0
[pip3] optree==0.19.1
[pip3] tbb==2023.0.0
[pip3] tbb-devel==2023.0.0
[pip3] tcmlib==1.5.0
[pip3] torch==2.12.0a0+git0d62256
[pip3] umf==1.1.0
[conda] intel-cmplr-lib-ur         2026.0.0             pypi_0           pypi
[conda] intel-openmp               2026.0.0             pypi_0           pypi
[conda] mkl                        2025.0.0             hacee8c2_941
[conda] mkl-include                2026.0.0             pypi_0           pypi
[conda] mkl-static                 2026.0.0             pypi_0           pypi
[conda] numpy                      1.26.4               pypi_0           pypi
[conda] onemkl-license             2026.0.0             pypi_0           pypi
[conda] optree                     0.19.1               pypi_0           pypi
[conda] tbb                        2023.0.0             pypi_0           pypi
[conda] tbb-devel                  2023.0.0             pypi_0           pypi
[conda] tcmlib                     1.5.0                pypi_0           pypi
[conda] torch                      2.12.0a0+git0d62256  pypi_0           pypi
[conda] umf                        1.1.0                pypi_0           pypi

cc @jbschlosser @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### PR DESCRIPTION
**Pull request by Charly21r**
## Summary

Calling the C++ API `at::batch_norm` (or `torch.batch_norm`) with a fewer-than-2-D input aborted the process with an internal `SmallVector` assertion instead of raising a user-facing error. This adds an input rank check so invalid input is rejected cleanly.

Fixes #186212 

## Root cause

Batch norm requires the channel dimension at index 1, i.e. an input of rank >= 2 `(N, C, ...)`. The public entry point never validated this:
- `_batch_norm_impl_index` reads `input.sym_sizes()[1]`. For a 1-D input this is out of bounds, but `ArrayRef::operator[]` does not bounds-check in release builds, so it silently read garbage. With `weight`/`bias`/ `running_*` left as `nullopt`, the `check_dims_match_num_input_features` validations were skipped and the bad value went unnoticed.
- Execution then reached `batch_norm_cpu`, where `ndim == 1` makes `DimVector reduce_dims(ndim - 1)` zero-length, and `reduce_dims[0] = 0;` indexes element 0 of a size-0 `SmallVector`. `SmallVector::operator[]` asserts `idx < size()`, which calls `abort()`.

The assertion was only a symptom; the defect was the missing rank check at the public entry point.

## Fix

Add `TORCH_CHECK(input.dim() >= 2, ...)` in `_batch_norm_impl_index`, before the `input.sym_sizes()[1]` access. This is the single dispatch choke point used by `at::batch_norm` for all backends (CPU/CUDA/cuDNN/MIOpen), so the check is device-independent and runs before backend selection. It mirrors the existing rank check in `renorm` in the same file.

Invalid input now raises a catchable `c10::Error` / Python `RuntimeError`:
```
batch_norm: expected input to have at least 2 dimensions (N, C, ...), but got input of dimension 1
```

## Tests

Added `test_batchnorm_raises_error_if_input_has_fewer_than_two_dims` in `test/test_nn.py`, alongside the existing `test_batchnorm_raises_error_*` cases. It calls the low-level `torch.batch_norm` op (which maps directly to `at::batch_norm`, the API from the bug report) with a 1-D input and the same all `None` weight/bias/running_* arguments, and asserts that a `RuntimeError` matching "at least 2 dimensions" is raised.

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
